pkg/kubelet/lifecycle/predicate.go

```diff
@@ -230,6 +230,17 @@ func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulerframework.No
 			podCopy.Spec.Containers[i].Resources.Requests[rName] = rQuant
 		}
 	}
+	for i, c := range pod.Spec.InitContainers {
+		podCopy.Spec.InitContainers[i].Resources.Requests = make(v1.ResourceList)
+		for rName, rQuant := range c.Resources.Requests {
+			if v1helper.IsExtendedResourceName(rName) {
+				if _, found := nodeInfo.Allocatable.ScalarResources[rName]; !found {
+					continue
+				}
+			}
+			podCopy.Spec.InitContainers[i].Resources.Requests[rName] = rQuant
+		}
+	}
```

Comment:
This code looks almost identical to the code above. Would it be possible to avoid duplication somehow?

Reply:
I’ll extract the duplicate code into a new function. Should I add this code to a new commit or just amend the existing one?

Reply:
I would propose this to not replicate the code:
```
func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulerframework.NodeInfo) *v1.Pod {
	podCopy := pod.DeepCopy()
	updateResourceRequests := func(containers []v1.Container) []v1.Container {
		for i, container := range containers {
			// We only handle requests in Requests but not Limits because the
			// PodFitsResources predicate, to which the result pod will be passed,
			// does not use Limits.
			container.Resources.Requests = filterExtendedResources(container.Resources.Requests, nodeInfo)
			containers[i] = container
		}
		return containers
	}

	podCopy.Spec.Containers = updateResourceRequests(podCopy.Spec.Containers)
	podCopy.Spec.InitContainers = updateResourceRequests(podCopy.Spec.InitContainers)

	return podCopy
}

func filterExtendedResources(requests v1.ResourceList, nodeInfo *schedulerframework.NodeInfo) v1.ResourceList {
	filteredRequests := make(v1.ResourceList)
	for rName, rQuant := range requests {
		if v1helper.IsExtendedResourceName(rName) {
			if _, found := nodeInfo.Allocatable.ScalarResources[rName]; !found {
				continue
			}
		}
		filteredRequests[rName] = rQuant
	}
	return filteredRequests
}
```

Reply:
nice suggestion, I plan to modify it like this:
```go
func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulerframework.NodeInfo) *v1.Pod {
	filterExtendedResources := func(containers []v1.Container) {
		for i, c := range containers {
			// We only handle requests in Requests but not Limits because the
			// PodFitsResources predicate, to which the result pod will be passed,
			// does not use Limits.
			filteredResources := make(v1.ResourceList)
			for rName, rQuant := range c.Resources.Requests {
				if v1helper.IsExtendedResourceName(rName) {
					if _, found := nodeInfo.Allocatable.ScalarResources[rName]; !found {
						continue
					}
				}
				filteredResources[rName] = rQuant
			}
			containers[i].Resources.Requests = filteredResources
		}
	}
	podCopy := pod.DeepCopy()
	filterExtendedResources(podCopy.Spec.Containers)
	filterExtendedResources(podCopy.Spec.InitContainers)
	return podCopy
}
```

Reply:
/lgtm